### PR TITLE
Migrate project to Deno with npm compatibility

### PR DESCRIPTION
**Pull request by alqubo**
- [x] Add .ts extension on import files
- [x] Fix errors with types
- [x] Added script for compile to npm with [dnt](https://github.com/denoland/dnt)
- [x] Changed test and publish pipelines 
- [x] Update readme
- [x] Change game refs to engine refs 
- [x] Changed tests to work with Deno
  ~~⚠️ Need check support for `not.toBeCalledWith` on Deno~~
  ~~⚠️ Need check test `expect entity to be updated only on system B` is not working~~
- [ ] Pipelines tested

Any suggestion of changes is welcome

### CODE REVIEW

**Comment by alqubo** [comment]
All set, except the pipelines tests.
@pagoru all yours, if you have any questions, tell me!

File: .github/workflows/publish.yml [hunk 1]
```diff
@@ -17,25 +17,33 @@ jobs:
...
+
       # Setup .npmrc file to publish to npm
       - uses: actions/setup-node@v3
         with:
           node-version: '16.x'
           registry-url: 'https://registry.npmjs.org'
       - name: Publish
-        run: npm publish
+        run: |
+          cd npm
```
Comment (pagoru) [inline .github/workflows/publish.yml:46]:
```suggestion
          rm npm.ts
          cd npm
```

File: src/engine.ts [hunk 2]
```diff
@@ -0,0 +1,382 @@
...
+  const _entity_getComponent = (
+    entityId: number,
+    component: any,
+    deepClone: boolean = false,
+  ) => {
+    const entityData = entityDataMap.get(entityId);
+    return entityData && entityData[component]
+      ? (
+        deepClone
+          ? JSON.parse(JSON.stringify(entityData[component]))
```
Comment (pagoru) [inline src/engine.ts:162]:
```suggestion
          ? structuredClone(entityData[component])
```

File: src/engine.ts [hunk 3]
```diff
@@ -0,0 +1,382 @@
...
+          : { ...entityData[component] }
+      )
+      : {};
+  };
+
+  const _entity_hasComponent = (entityId: number, component: any) =>
+    entityComponentMap[entityId]?.includes(component);
+
+  const _entity_getData = (entityId: number) =>
+    JSON.parse(JSON.stringify(entityDataMap.get(entityId)));
```
Comment (pagoru) [inline src/engine.ts:172]:
```suggestion
   structuredClone(entityDataMap.get(entityId));
```

File: src/types.ts [hunk 4]
```diff
@@ -1,42 +1,41 @@
...
-    
-    getEntityList?: () => EntityType[],
-    getEntityListByType?: (type: number) => EntityType[],
-    getEntityListByComponents?: (...componentList: string[]) => EntityType[],
+export interface EngineType {
+  setSystems: (...systems: SystemFunction[]) => any;
+
+  getEntityList: () => EntityType[];
+  getEntityListByType: (type: number) => EntityType[];
+  getEntityListByComponents: (...componentList: string[]) => EntityType[];
```
Comment (pagoru) [inline src/types.ts:9]:
```suggestion
  getEntityListByComponents: (...componentList: unknown[]) => EntityType[];
```

File: src/types.ts [hunk 5]
```diff
@@ -1,42 +1,41 @@
...
-    id?: string;
-    components: string[];
-    onAdd?: (id: number) => void;
-    onUpdate?: (id: number, component?: string) => void;
-    onRemove?: (id: number) => void;
-    
-    onLoad?: () => void;
-    onDestroy?: () => void;
+  id: string;
+  components: string[];
```
Comment (pagoru) [inline src/types.ts:32]:
```suggestion
  components: unknown[];
```

File: src/types.ts [hunk 6]
```diff
@@ -45,22 +44,31 @@ export type SystemFunction = () => SystemType;
...
-    readonly shortcuts?: Record<string, <T>(entity: EntityType, data?: T) => unknown>;
+  //Only initial declaration
+  readonly id: number;
+  readonly type: number;
+  readonly data: Record<string, Object>;
+  readonly components: string[];
+  readonly shortcuts?: Record<
+    string,
+    <T>(entity: EntityType, data?: T) => unknown
+  >;
```
Comment (pagoru) [inline src/types.ts:55]:
```suggestion
  readonly data: Record<unknown, Object>;
  readonly components: unknown[];
```

File: src/types.ts [hunk 7]
```diff
@@ -45,22 +44,31 @@ export type SystemFunction = () => SystemType;
...
+  getComponent?: <ComponentType>(
+    component: string,
+    deepClone?: boolean,
+  ) => ComponentType;
+  getComponents?: () => string[];
+  hasComponent?: (component: string) => boolean;
+  updateComponent?: UpdateComponentFunctionType;
+  removeComponent?: RemoveComponentFunctionType;
+  // shortcuts
+  actions?: Record<string, <T>(data?: T) => unknown>;
```
Comment (pagoru) [inline src/types.ts:67]:
```suggestion
  getData?: () => Record<unknown, Object>;
  getComponent?: <ComponentType>(
    component: unknown,
    deepClone?: boolean,
  ) => ComponentType;
  getComponents?: () => unknown[];
  hasComponent?: (component: unknown) => boolean;
  updateComponent?: UpdateComponentFunctionType;
  removeComponent?: RemoveComponentFunctionType;
```

**Review by pagoru — APPROVED** [review]

**Pull request merged**